### PR TITLE
feat: rewrite Checkbox as a separate module | suit to Fiori3

### LINKED ISSUE #1769: Checkbox 0.4.0 adaptation
On fundamental-styles 0.4.0 there is changed structure of checkbox. We should have another component for this.

### PR DESCRIPTION
**Pull request by salarenko**
#### Please provide a link to the associated issue.
closes #1769 

#### Please provide a brief summary of this pull request.
This pull requests includes:
- new implementation of checkbox control as separate component and module
- new checkbox component supporting Fiori3 specification
- documentation to new component

Checkbox state is driven by value passed to component by FormControl.

#### Please check whether the PR fulfills the following requirements
- [x] tests for the changes that have been done

Documentation checklist:
- [x] Documentation Examples
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)

### CODE REVIEW

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    get isChecked(): boolean {
+        return this.checkboxState === 'checked';
+    }
+
+    get isIndeterminate(): boolean {
+        return this.checkboxState === 'indeterminate';
+    }
+
+    writeValue(value: any): void {
```
Comment (JKMarkowski) [inline libs/core/src/lib/checkbox/checkbox/checkbox.component.ts]: Let's mark methods, that are implemented from `ControlValueAccessor` as `/** @hidden */`, they shouldn't be in documentation IMO. Let's shortly describe all of other methods.

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts [hunk 2]
```diff
@@ -1,42 +1,60 @@
...
+        }
+    ];
+
+    checkboxCustomValues: ExampleFile[] = [
         {
-            language: 'typescript',
-            code: formGroupInputTs,
-            fileName: 'checkbox-form-group-example',
-            component: 'CheckboxFormGroupExampleComponent'
+            language: 'TypeScript',
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: typescript should have lower case 't' for stackblitz to work

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts [hunk 3]
```diff
@@ -1,42 +1,60 @@
...
+            fileName: 'checkbox-custom-values-example',
+            component: 'CheckboxCustomValuesExampleComponent',
+            code: checkboxCustomValuesTsCode
         }
     ];
 
-    ngOnInit() { }
+    checkboxReactiveForms: ExampleFile[] = [
+        {
+            language: 'TypeScript',
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: same here

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts [hunk 4]
```diff
@@ -1,42 +1,60 @@
...
+            fileName: 'checkbox-reactive-forms-example',
+            component: 'CheckboxReactiveFormsExampleComponent',
+            code: checkboxReactiveFormsTsCode
+        }
+    ];
+
+    checkboxStates: ExampleFile[] = [
+        {
+            language: 'TypeScript',
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: same here

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts [hunk 5]
```diff
@@ -1,42 +1,60 @@
...
-        },
+            language: 'TypeScript',
+            fileName: 'checkbox-tristate-example',
+            component: 'DefaultCheckboxExamplesComponent',
+            code: checkboxTristateTsCode
+        }
+    ];
+
+    checkboxCustomValues: ExampleFile[] = [
         {
```
Comment (stefanoScalzo) [inline apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: You need to add the html as well for all

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        fixture.detectChanges();
+
+        hostComponent.value = false;
+        fixture.detectChanges();
+
+        expect(checkbox.writeValue).toHaveBeenCalledTimes(3);
+        expect(checkbox.checkboxValue).toBe(false);
+    });
+
+    // it('should be unchecked on double click', () => {
```
Comment (stefanoScalzo) [inline libs/core/src/lib/checkbox/checkbox/checkbox.component.spec.ts]: What are all of these comments for. IF they are necessary remove the ones with nothing inside

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    @Input()
+    tristate: boolean;
+
+    /** Allows to prevent user from manually selecting controls third state. */
+    @Input()
+    tristateSelectable: boolean = true;
+
+    /** Values returned by control. */
+    @Input()
+    values: FdCheckboxValues = {trueValue: true, falseValue: false, thirdStateValue: null};
```
Comment (stefanoScalzo) [inline libs/core/src/lib/checkbox/checkbox/checkbox.component.ts]: I know by doing this we are giving them flexibility but shouldn't a checkbox be strictly a boolean ?

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /** Based on current control state:
+     * - sets next control value
+     * - emits new control value
+     * - updates control state based on new control value
+     * */
+    public nextValue(): void {
+
+        if (this.disabled) {
```
Comment (stefanoScalzo) [inline libs/core/src/lib/checkbox/checkbox/checkbox.component.ts]: is this necessary?

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.ts (hunk 1 above)
Comment (salarenko) [reply libs/core/src/lib/checkbox/checkbox/checkbox.component.ts]: Done

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts (hunk 2 above)
Comment (salarenko) [reply apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: Done

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts (hunk 3 above)
Comment (salarenko) [reply apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: Done

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts (hunk 4 above)
Comment (salarenko) [reply apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: Done

File: apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts (hunk 5 above)
Comment (salarenko) [reply apps/docs/src/app/core/component-docs/checkbox/checkbox-docs.component.ts]: Done

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.spec.ts (hunk 6 above)
Comment (salarenko) [reply libs/core/src/lib/checkbox/checkbox/checkbox.component.spec.ts]: Test cases have been updated

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.ts (hunk 8 above)
Comment (salarenko) [reply libs/core/src/lib/checkbox/checkbox/checkbox.component.ts]: Removed

File: libs/core/src/lib/checkbox/checkbox/checkbox.component.ts (hunk 7 above)
Comment (salarenko) [reply libs/core/src/lib/checkbox/checkbox/checkbox.component.ts]: Checkbox control returns boolean values by default.
I think giving user possibility to define values returned by checkbox is a nice and useful feature.

**Comment by InnaAtanasova** [comment]
> Looks very well for me. Great Job, the `compare-object` function will be useful in feature..

I complete agree with Jedrzej . Great job!

**Comment by InnaAtanasova** [comment]
The build is failing. Could you take a look at it?

**Pull request merged**